editor/src/core/workers/parser-printer/uid-fix.ts

```diff
@@ -1,284 +1,569 @@
-      return true
-    } else if (isJSXConditionalExpression(oldElement) && isJSXConditionalExpression(newElement)) {
-      const oldUID = getUtopiaID(oldElement)
-      const newUid = getUtopiaID(newElement)
-      const path = EP.appendToElementPath(pathSoFar, newUid)
-      const oldPathToRestore = EP.appendToElementPath(pathSoFar, oldUID)
-      onElement(oldUID, newUid, oldPathToRestore, path)
-      const whenTrue = compareAndWalkElements(
-        oldElement.whenTrue,
-        newElement.whenTrue,
-        path,
-        onElement,
-      )
-      const whenFalse = compareAndWalkElements(
-        oldElement.whenFalse,
-        newElement.whenFalse,
-        path,
-        onElement,
-      )
-      return whenTrue && whenFalse
+  fixUIDsState: FixUIDsState,
+): JSXElementChild {
+  switch (newElement.type) {
+    case 'JSX_ELEMENT': {
+      if (oldElement.type === newElement.type) {
+        return fixJSXElementUIDs(oldElement, newElement, fixUIDsState)
+      }
+      break
+    }
+    case 'JSX_FRAGMENT': {
+      if (oldElement.type === newElement.type) {
+        const updatedChildren = fixJSXElementChildArray(
+          oldElement.children,
+          newElement.children,
+          fixUIDsState,
+        )
+        return updateUID(jsxFragmentUIDOptic, oldElement.uid, fixUIDsState, {
+          ...newElement,
+          children: updatedChildren,
+        })
+      }
+      break
+    }
+    case 'JSX_TEXT_BLOCK': {
+      if (oldElement.type === newElement.type) {
+        return updateUID(jsxTextBlockUIDOptic, oldElement.uid, fixUIDsState, newElement)
+      }
+      break
+    }
+    case 'JSX_CONDITIONAL_EXPRESSION': {
+      if (oldElement.type === newElement.type) {
+        const updatedCondition = fixExpressionUIDs(
+          oldElement.condition,
+          newElement.condition,
+          fixUIDsState,
+        )
+        const updatedWhenTrue = fixJSXElementChildUIDs(
+          oldElement.whenTrue,
+          newElement.whenTrue,
+          fixUIDsState,
+        )
+        const updatedWhenFalse = fixJSXElementChildUIDs(
+          oldElement.whenFalse,
+          newElement.whenFalse,
+          fixUIDsState,
+        )
+        return updateUID(jsxConditionalExpressionUIDOptic, oldElement.uid, fixUIDsState, {
+          ...newElement,
+          condition: updatedCondition,
+          whenTrue: updatedWhenTrue,
+          whenFalse: updatedWhenFalse,
+        })
+      }
+      break
+    }
+    case 'ATTRIBUTE_VALUE':
+    case 'ATTRIBUTE_NESTED_ARRAY':
+    case 'ATTRIBUTE_NESTED_OBJECT':
+    case 'ATTRIBUTE_FUNCTION_CALL':
+    case 'ATTRIBUTE_OTHER_JAVASCRIPT': {
+      if (oldElement.type === newElement.type) {
+        return fixExpressionUIDs(oldElement, newElement, fixUIDsState)
+      }
+      break
+    }
+    default:
+      const _exhaustiveCheck: never = newElement
+      throw new Error(`Unhandled case ${JSON.stringify(newElement)}.`)
```

Comment:
can this also be replaced with `assertNever`?

Reply:
Fixed in 6a3d04e93.